crypto/monero/address.go

```diff
@@ -20,83 +46,141 @@ var (
 	errInvalidAddressEncoding   = errors.New("invalid monero address encoding")
 	errInvalidPrefixGotMainnet  = errors.New("invalid monero address: expected stagenet, got mainnet")
 	errInvalidPrefixGotStagenet = errors.New("invalid monero address: expected mainnet, got stagenet")
+	errInvalidPrefixGotTestnet  = errors.New("invalid monero address: monero testnet not yet supported")
 )
 
-// Address represents a base58-encoded string
-type Address string
+// Address represents a Monero address
+type Address struct {
+	// decoded is the bytes (prefix, pub spend key, pub view key, checksum) that
+	// get base58 encoded. Package private, as it is a semi-arbitrary
+	// implementation detail.
+	decoded [addressBytesLen]byte
+}
 
-// ValidateAddress checks if the given address is valid
-func ValidateAddress(addr string, env common.Environment) error {
-	b, err := MoneroAddrBase58ToBytes(addr)
-	if err != nil {
-		return err
+// NewAddress converts a string to a monero Address with validation.
+func NewAddress(addrStr string, env common.Environment) (*Address, error) {
+	addr := new(Address)
+	if err := addr.UnmarshalText([]byte(addrStr)); err != nil {
+		return nil, err
 	}
 
-	switch env {
-	case common.Mainnet, common.Development:
-		if b[0] != addressPrefixMainnet {
-			return errInvalidPrefixGotStagenet
-		}
-	case common.Stagenet:
-		if b[0] != addressPrefixStagenet {
-			return errInvalidPrefixGotMainnet
-		}
+	return addr, addr.ValidateEnv(env)
+}
+
+func (a *Address) String() string {
+	return addrBytesToBase58(a.decoded[:])
+}
+
+// Network returns the Monero network of the address
+func (a *Address) Network() Network {
+	switch a.decoded[0] {
+	case netPrefixStdAddrMainnet, netPrefixSubAddrMainnet:
+		return Mainnet
+	case netPrefixStdAddrStagenet, netPrefixSubAddrStagenet:
+		return Stagenet
+	case netPrefixStdAddrTestnet, netPrefixSubAddrTestnet:
+		return Testnet
+	default:
+		// Our methods to deserialize and create Address values all verify
+		// that the address byte is valid
+		panic("address has invalid network prefix")
+	}
+}
+
+// Type returns the Address type
+func (a *Address) Type() AddressType {
```

Comment:
I'm going to leave this here, but it doesn't actually appear to be needed at the current time. I thought it might be needed when the XMR Taker receives the NotifyXMRLock message. When that happens, I thought we might need to validate that the passed address is standard/primary, but that code is already comparing the received address with a value it generated from keys, which already guarantees that the address is standard/primary.